### PR TITLE
ARROW-16773: [Docs][Format] Document Run-Length encoding in Arrow columnar format

### PR DESCRIPTION
**Pull request by zagto**

### CODE REVIEW

File: docs/source/format/Columnar.rst [hunk 1]
```diff
@@ -765,6 +765,85 @@ application.
 We discuss dictionary encoding as it relates to serialization further
 below.
 
+.. _run-length-encoded-layout:
+
+Run-Length-encoded Layout
+-------------------------
+
+Run-Length is a data representation that represents data as sequences of the
+same a, called runs. Each run is represented as a value, and an integer
```
Comment (lidavidm) [inline docs/source/format/Columnar.rst:774]: "same value"?

File: docs/source/format/Columnar.rst [hunk 2]
```diff
@@ -765,6 +765,85 @@ application.
...
+describing how often this value is repeated.
+
+Any array can be run-length-encoded. A run-length encoded array has a single
+buffer holding as many 32-bit integers, as there are runs. The actual values are
+hold in a child array, which is just a regular array
```
Comment (lidavidm) [inline docs/source/format/Columnar.rst:779]:
```suggestion
held in a child array, which is just a regular array.
```

File: docs/source/format/Columnar.rst [hunk 3]
```diff
@@ -765,6 +765,85 @@ application.
...
+
+The dictionary is stored as an optional
+property of an array. When a field is dictionary encoded, the values are
+represented by an array of non-negative integers representing the index of the
+value in the dictionary. The memory layout for a dictionary-encoded array is
+the same as that of a primitive integer layout. The dictionary is handled as a
+separate columnar array with its own respective layout.
```
Comment (lidavidm) [inline docs/source/format/Columnar.rst:786]: Is this related? Or was this a copy-paste thing

File: docs/source/format/Columnar.rst [hunk 4]
```diff
@@ -765,6 +765,85 @@ application.
...
+
+    type: Float32
+
+    [1.0, 1.0, 1.0, 1.0, null, null, 2.0]
+
+In Run-length-encoded form, this could appear as:
+
+::
+
+    * Length: 3, Null count: 2
```
Comment (lidavidm) [inline docs/source/format/Columnar.rst:798]: It is a bit odd to see both logical (length) and physical (null count) values at the top level, it seems inconsistent

File: docs/source/format/Columnar.rst [hunk 5]
```diff
@@ -765,6 +765,85 @@ application.
...
+    * Accumulated run lengths buffer:
+
+      | Bytes 0-3   | Bytes 4-7   | Bytes 8-11  | Bytes  6-63           |
+      |-------------|-------------|-------------|-----------------------|
+      | 4           | 6           | 7           | unspecified (padding) |
```
Comment (lidavidm) [inline docs/source/format/Columnar.rst:803]: Shouldn't these be 4, 2, 1?

Comment (lidavidm) [reply docs/source/format/Columnar.rst:803]: How are top-level nulls interpreted? Or are they disallowed?

File: docs/source/format/Columnar.rst [hunk 6]
```diff
@@ -765,6 +765,85 @@ application.
...
+of its indices, irrespective of any null values in the dictionary.
+
+Since unsigned integers can be more difficult to work with in some cases
+(e.g. in the JVM), we recommend preferring signed integers over unsigned
+integers for representing dictionary indices. Additionally, we recommend
+avoiding using 64-bit unsigned integer indices unless they are required by an
+application.
+
+We discuss dictionary encoding as it relates to serialization further
+below.
```
Comment (lidavidm) [inline docs/source/format/Columnar.rst:845]: same here, was this meant to be kept?

File: docs/source/format/Columnar.rst [hunk 7]
```diff
@@ -765,6 +765,85 @@ application.
...
+
+Run-Length-encoded Layout
+-------------------------
+
+Run-Length is a data representation that represents data as sequences of the
+same a, called runs. Each run is represented as a value, and an integer
+describing how often this value is repeated.
+
+Any array can be run-length-encoded. A run-length encoded array has a single
+buffer holding as many 32-bit integers, as there are runs. The actual values are
```
Comment (lidavidm) [inline docs/source/format/Columnar.rst:778]: nit: explicitly say signed integers?

File: docs/source/format/Columnar.rst (hunk 5 above)
Comment (zagto) [reply docs/source/format/Columnar.rst:803]: > How are top-level nulls interpreted? Or are they disallowed?

The idea for now is that the top level only has a single buffer, holding the run length values. So it is not possible to a have a null three

> Shouldn't these be 4, 2, 1?

No they hold the accumulated length of all runs from the first to the current one. The paragraph above should have said that, sorry.

The reason for storing accumulated lengths is to make random access to a logical index possible using binary search. We can still get the length of the current run efficiently by subtracting the two adjacent values

File: docs/source/format/Columnar.rst (hunk 3 above)
Comment (zagto) [reply docs/source/format/Columnar.rst:786]: copy-paste mistake indeed

File: docs/source/format/Columnar.rst (hunk 5 above)
Comment (lidavidm) [reply docs/source/format/Columnar.rst:803]: Oops, right, you've mentioned that. Thanks!

File: docs/source/format/Columnar.rst (hunk 4 above)
Comment (zagto) [reply docs/source/format/Columnar.rst:798]: true. one more reason I should probably change the proposal to store the logical length there

File: docs/source/format/Columnar.rst (hunk 5 above)
Comment (zagto) [reply docs/source/format/Columnar.rst:808]: I fixed this and the other problems you commented. Thanks for looking into this!

File: docs/source/format/Columnar.rst [hunk 8]
```diff
@@ -765,6 +765,65 @@ application.
...
+
+Run-Length-encoded Layout
+-------------------------
+
+Run-Length is a data representation that represents data as sequences of the
+same value, called runs. Each run is represented as a value, and an integer
+describing how often this value is repeated.
+
+Any array can be run-length-encoded. A run-length encoded array has a single
+buffer holding as many signed 32-bit integers, as there are runs. The actual
```
Comment (zagto) [inline docs/source/format/Columnar.rst:778]: This is actually not true if we use 32 bits. An array whose length is larger than a 32 bit integer cannot be represented

File: docs/source/format/Columnar.rst [hunk 9]
```diff
@@ -765,6 +765,65 @@ application.
 We discuss dictionary encoding as it relates to serialization further
 below.
 
+.. _run-length-encoded-layout:
+
+Run-Length-encoded Layout
```
Comment (westonpace) [inline docs/source/format/Columnar.rst:770]:
```suggestion
Run-Length Encoded Layout
```

File: docs/source/format/Columnar.rst [hunk 10]
```diff
@@ -765,6 +765,65 @@ application.
...
+-------------------------
+
+Run-Length is a data representation that represents data as sequences of the
+same value, called runs. Each run is represented as a value, and an integer
+describing how often this value is repeated.
+
+Any array can be run-length-encoded. A run-length encoded array has a single
```
Comment (westonpace) [inline docs/source/format/Columnar.rst:777]:
```suggestion
Any array can be run-length encoded. A run-length encoded array has a single
```

File: docs/source/format/Columnar.rst (hunk 8 above)
Comment (westonpace) [inline docs/source/format/Columnar.rst:778]:
```suggestion
buffer holding a signed 32-bit integer for each run. The actual
```

File: docs/source/format/Columnar.rst [hunk 11]
```diff
@@ -957,7 +1016,7 @@ The ``Buffer`` Flatbuffers value describes the location and size of a
 piece of memory. Generally these are interpreted relative to the
 **encapsulated message format** defined below.
 
-The ``size`` field of ``Buffer`` is not required to account for padding
+The ``size`` field of ``Buffer`` is not required to account for paddingeng-career-mgmt
```
Comment (westonpace) [inline docs/source/format/Columnar.rst:1019]:
```suggestion
The ``size`` field of ``Buffer`` is not required to account for padding
```

**Review by westonpace — COMMENTED** [review]
Thanks, this is very readable.

**Comment by lidavidm** [comment]
This looks reasonable to me, thanks.

File: docs/source/format/Columnar.rst [hunk 12]
```diff
@@ -765,6 +765,66 @@ application.
...
+
+Run-Length Encoded Layout
+-------------------------
+
+Run-Length is a data representation that represents data as sequences of the
+same value, called runs. Each run is represented as a value, and an integer
+describing how often this value is repeated.
+
+Any array can be run-length encoded. A run-length encoded array has no buffers
+by itself, but has two child arrays. The first one holds a signed 32-bit integer
```
Comment (alamb) [inline docs/source/format/Columnar.rst:778]:
```suggestion
by itself, but has two child arrays. The first one holds a signed 32-bit integer called a "run end"
```

File: docs/source/format/Columnar.rst [hunk 13]
```diff
@@ -765,6 +765,66 @@ application.
...
+for each run. The actual values of each run are held the second child array.
+
+The values in the first child array represent the length of each run. They do
+not hold the length of the respective run directly, but the accumulated length
+of all runs from the first to the current one, i.e. the logical index where the
+current run ends. This allows relatively efficient random access from a logical
+index using binary search. The length of an individual run can be determined by
+subtracting two adjacent values.
+
+A run has to have a length of at least 1. This means the values in the
```
Comment (alamb) [inline docs/source/format/Columnar.rst:788]:
```suggestion
A run must have have a length of at least 1. This means the values in the
```

File: docs/source/format/Columnar.rst [hunk 14]
```diff
@@ -765,6 +765,66 @@ application.
...
+run ends array all positive and in strictly ascending order. A run end cannot be
+null.
```
Comment (alamb) [inline docs/source/format/Columnar.rst:790]: I recommend adding something to this section saying "RLE Arrays must not have have a validity bitmap" which would also imply a run end could not be null. As worded I think it is unclear if a validity bitmap is allowed but must be all `valid` or if the bitmap is not allowed

File: docs/source/format/Columnar.rst [hunk 15]
```diff
@@ -765,6 +765,66 @@ application.
...
+In Run-length-encoded form, this could appear as:
+
+::
+
+    * Length: 7, Null count: 2
+    * Children arrays:
+
+      * run ends (Int32):
+        * Length: 3, Null count: 0
+        * Validity bitmap buffer: Not required
```
Comment (alamb) [inline docs/source/format/Columnar.rst:806]:
```suggestion
        * Validity bitmap buffer: Not present (not allowed)
```
See comment above

**Review by alamb — COMMENTED** [review]
Looks good to me. Thank you for this @zagto

**Comment by alamb** [comment]
FYI @tustvold @viirya @nevi-me in case you are interested in this proposed format addition

File: docs/source/format/Columnar.rst (hunk 14 above)
Comment (alamb) [reply docs/source/format/Columnar.rst:792]: " A run end cannot be null." sounds good to me

File: docs/source/format/Columnar.rst (hunk 15 above)
Comment (alamb) [reply docs/source/format/Columnar.rst:808]: I still think it would be best to be explicit here that no validity bitmap is allowed to avoid ambiguity

**Comment by zeroshade** [comment]
Closing as this has been incorporated into #14176

File: docs/source/format/Columnar.rst [hunk 16]
```diff
@@ -765,6 +765,68 @@ application.
...
+As an example, you could have the following data: ::
+
+    type: Float32
+    [1.0, 1.0, 1.0, 1.0, null, null, 2.0]
+
+In Run-length-encoded form, this could appear as:
+
+::
+
+    * Length: 7, Null count: 2
```
Comment (tustvold) [inline docs/source/format/Columnar.rst:803]: This formulation of null count is a little surprising, it implies that when computing the parent ArrayData from a set of children, it must iterate the null mask in conjunction with the run ends. This is also inconsistent with how null counts work for other nested types

Comment (zagto) [reply docs/source/format/Columnar.rst:803]: You're right, it should actually say 0, since the array has no validity bitmap

(I tried to reply via the reply button in the ml web ui but that does not seem to have come through)

Comment (tustvold) [reply docs/source/format/Columnar.rst:803]: > since the array has no validity bitmap

Is this a general restriction, or is it just in this case there is no validity mask?

Comment (zagto) [reply docs/source/format/Columnar.rst:803]: Yes, this is a general restriction (at least I think so, and that's how the prototype code works). The idea so far is that Null is just one of the possible values for a run.

The is somewhat consistent with how union types work.

(if we were to allow the RLE array parent to have an additional null mask, the null count field would represent that - there seems to be a generall assumption in Arrow code that a non-zero (or array length for the NULL) null count means the presence of the standard null mask)

Comment (tustvold) [reply docs/source/format/Columnar.rst:803]: > this is a general restriction

We should probably document that if so, personally it seems a shame as it complicates things like nullif kernels which now have to explicitly handle the case of RLE data, whereas normally they don't need concern themselves with anything but the array data.

> union types work

Yeah, union types are a bit of a special snowflake though :laughing:

Comment (zagto) [reply docs/source/format/Columnar.rst:803]: it is actually documented by saying there are no buffers, but that's a bit indirect, so guess mentioning it explicitly won't hurt

Comment (zagto) [reply docs/source/format/Columnar.rst:803]: I submitted a PR for this here, please take a look:
[https://github.com/apache/arrow/pull/33831](https://github.com/apache/arrow/pull/33831)
